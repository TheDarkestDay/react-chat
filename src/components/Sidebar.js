import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import Input from 'material-ui/Input';
import Paper from 'material-ui/Paper';
import List, { ListItem, ListItemText } from 'material-ui/List';
import Avatar from 'material-ui/Avatar';
import Button from 'material-ui/Button';
import AddIcon from 'material-ui-icons/Add';
import BottomNavigation, { BottomNavigationAction } from 'material-ui/BottomNavigation';
import RestoreIcon from 'material-ui-icons/Restore';
import ExploreIcon from 'material-ui-icons/Explore';
import toMaterialStyle from 'material-color-hash';
import { withStyles } from 'material-ui/styles';

import { Chat } from '../constants/shapes';
import generateLastActivityMessage from '../services/generate-last-activity-message';
import CreateChannelDialog from './dialogs/CreateChannelDialog';

const chatSearchFormHeight = '64px';
const bottomNavigationHeight = '56px';

const styles = () => ({
  chatLink: {
    textDecoration: 'none',
  },
  searchChatForm: {
    padding: '16px 24px',
  },
  sidebar: {
    position: 'relative',
    display: 'flex',
    flexDirection: 'column',
    height: '100%',
    backgroundColor: '#ffffff',
  },
  chatsList: {
    flexGrow: 1,
    overflow: 'auto',
    maxHeight: `calc(100vh - ${chatSearchFormHeight} - ${bottomNavigationHeight})`,
  },
  addChatBtn: {
    position: 'absolute',
    bottom: '64px',
    right: '24px',
  },
});

const CHATS_SELECTION = {
  MY: 0,
  ALL: 1,
};

class Sidebar extends Component {
  static propTypes = {
    classes: PropTypes.objectOf(PropTypes.string).isRequired,
    chats: PropTypes.arrayOf(Chat).isRequired,
    chatQuery: PropTypes.string.isRequired,
    isSocketConnected: PropTypes.bool.isRequired,
    getChats: PropTypes.func.isRequired,
    showMyChats: PropTypes.func.isRequired,
    showAllChats: PropTypes.func.isRequired,
    searchChats: PropTypes.func.isRequired,
    createChat: PropTypes.func.isRequired,
  };

  state = {
    currentSelectionIndex: 0,
    isDialogOpened: false,
  };

  componentDidMount() {
    this.props.getChats();
    this.props.showMyChats();
  }

  handleChatQueryChange = (event) => {
    this.props.searchChats(event.target.value);
  };

  handleDialogClose = () => {
    this.setState({
      isDialogOpened: false,
    });
  };

  handleDialogDone = (chatTitle) => {
    this.handleDialogClose();
    this.props.createChat(chatTitle);
  };

  toggleChatsSelection = (event, value) => {
    this.setState({
      currentSelectionIndex: value,
    });

    if (value === CHATS_SELECTION.MY) {
      this.props.showMyChats();
    } else {
      this.props.showAllChats();
    }
  };

  openCreateChannelDialog = () => {
    this.setState({
      isDialogOpened: true,
    });
  };

  render() {
    const {
      classes, chats, chatQuery, isSocketConnected,
    } = this.props;
    const { currentSelectionIndex, isDialogOpened } = this.state;

    return (
      <aside className={classes.sidebar}>
        <Paper className={classes.searchChatForm} elevation={1}>
          <form>
            <Input
              id="chatSearch"
              placeholder="Search chats..."
              value={chatQuery}
              onChange={this.handleChatQueryChange}
              fullWidth
            />
          </form>
        </Paper>
        <List className={classes.chatsList}>
          {chats.map(chat => (
            <Link
              href={`/chat/${chat._id}`}
              className={classes.chatLink}
              to={`/chat/${chat._id}`}
              key={chat._id}
            >
              <ListItem button>
                <Avatar style={toMaterialStyle(chat.title)}>{chat.title[0]}</Avatar>
                <ListItemText
                  primary={chat.title}
                  secondary={generateLastActivityMessage(chat.updatedAt)}
                />
              </ListItem>
            </Link>
          ))}
        </List>
        <Button
          className={classes.addChatBtn}
          onClick={this.openCreateChannelDialog}
          variant="fab"
          color="primary"
          aria-label="add"
          disabled={!isSocketConnected}
        >
          <AddIcon />
        </Button>
        <BottomNavigation
          value={currentSelectionIndex}
          onChange={this.toggleChatsSelection}
          showLabels
        >
          <BottomNavigationAction label="My Chats" icon={<RestoreIcon />} />
          <BottomNavigationAction label="Explore" icon={<ExploreIcon />} />
        </BottomNavigation>
        <CreateChannelDialog
          isOpened={isDialogOpened}
          onClose={this.handleDialogClose}
          onDone={this.handleDialogDone}
          disabled={!isSocketConnected}
        />
      </aside>
    );
  }
}

export default withStyles(styles)(Sidebar);
