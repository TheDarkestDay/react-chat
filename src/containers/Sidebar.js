import { connect } from 'react-redux';

import { createChat, getChats, showAllChats, showMyChats } from '../actions/chat';
import { getVisibleChats } from '../selectors';
import Sidebar from '../components/Sidebar';

const mapStateToProps = (state) => {
  return {
    chats: getVisibleChats(state)
  }
};

const mapDispatchToProps = (dispatch) => {
  return {
    createChat: (chatData) => {
      dispatch(createChat(chatData));
    },
    getChats: () => {
      dispatch(getChats());
    },
    showAllChats: () => {
      dispatch(showAllChats());
    },
    showMyChats: () => {
      dispatch(showMyChats());
    }
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Sidebar);