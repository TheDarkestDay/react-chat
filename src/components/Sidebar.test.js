/* eslint-env jest */
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Sidebar from './Sidebar';

it('renders without crashing', () => {
  const sidebarPropsMock = {
    chats: [],
    chatQuery: '',
    isSocketConnected: true,
    getChats: () => {},
    showMyChats: () => {},
    showAllChats: () => {},
    searchChats: () => {},
    createChat: () => {},
  };

  const div = document.createElement('div');
  ReactDOM.render(<Sidebar {...sidebarPropsMock} />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('should match snapshot', () => {
  const sidebarPropsMock = {
    chats: [],
    chatQuery: '',
    isSocketConnected: true,
    getChats: () => {},
    showMyChats: () => {},
    showAllChats: () => {},
    searchChats: () => {},
    createChat: () => {},
  };

  const tree = renderer.create(<Sidebar {...sidebarPropsMock} />).toJSON();
  expect(tree).toMatchSnapshot();
});
