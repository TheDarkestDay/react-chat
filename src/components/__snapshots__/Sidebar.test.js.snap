// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<aside
  className="Sidebar-sidebar-102"
>
  <div
    className="MuiPaper-root-105 MuiPaper-shadow1-108 MuiPaper-rounded-106 Sidebar-searchChatForm-101"
  >
    <form>
      <div
        className="MuiInput-root-132 MuiInput-fullWidth-140 MuiInput-inkbar-134 MuiInput-underline-138"
        onBlur={[Function]}
        onFocus={[Function]}
      >
        <input
          aria-invalid={undefined}
          aria-required={undefined}
          autoComplete={undefined}
          autoFocus={undefined}
          className="MuiInput-input-141"
          defaultValue={undefined}
          disabled={undefined}
          id="chatSearch"
          name={undefined}
          onChange={[Function]}
          onKeyDown={undefined}
          onKeyUp={undefined}
          placeholder="Search chats..."
          readOnly={undefined}
          required={undefined}
          rows={undefined}
          type="text"
          value=""
        />
      </div>
    </form>
  </div>
  <ul
    className="MuiList-root-147 MuiList-padding-148 Sidebar-chatsList-103"
  />
  <button
    aria-label="add"
    className="MuiButtonBase-root-166 MuiButton-root-151 MuiButton-raised-156 MuiButton-fab-161 MuiButton-raisedPrimary-158 Sidebar-addChatBtn-104"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <span
      className="MuiButton-label-152"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-168 MuiSvgIcon-fontSize-174"
        color={undefined}
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
    </span>
    <span
      className="MuiTouchRipple-root-175"
    />
  </button>
  <div
    className="MuiBottomNavigation-root-182"
  >
    <button
      className="MuiButtonBase-root-166 MuiBottomNavigationAction-root-183 MuiBottomNavigationAction-selected-184"
      disabled={undefined}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiBottomNavigationAction-wrapper-186"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-168"
          color={undefined}
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M13 3c-4.97 0-9 4.03-9 9H1l3.89 3.89.07.14L9 12H6c0-3.87 3.13-7 7-7s7 3.13 7 7-3.13 7-7 7c-1.93 0-3.68-.79-4.94-2.06l-1.42 1.42C8.27 19.99 10.51 21 13 21c4.97 0 9-4.03 9-9s-4.03-9-9-9zm-1 5v5l4.28 2.54.72-1.21-3.5-2.08V8H12z"
          />
        </svg>
        <span
          className="MuiBottomNavigationAction-label-187 MuiBottomNavigationAction-selectedLabel-188"
        >
          My Chats
        </span>
      </span>
      <span
        className="MuiTouchRipple-root-175"
      />
    </button>
    <button
      className="MuiButtonBase-root-166 MuiBottomNavigationAction-root-183"
      disabled={undefined}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiBottomNavigationAction-wrapper-186"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-168"
          color={undefined}
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 10.9c-.61 0-1.1.49-1.1 1.1s.49 1.1 1.1 1.1c.61 0 1.1-.49 1.1-1.1s-.49-1.1-1.1-1.1zM12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm2.19 12.19L6 18l3.81-8.19L18 6l-3.81 8.19z"
          />
        </svg>
        <span
          className="MuiBottomNavigationAction-label-187"
        >
          Explore
        </span>
      </span>
      <span
        className="MuiTouchRipple-root-175"
      />
    </button>
  </div>
</aside>
`;
